Giving xp as a GM (targeting a character token) generates an error on the console

**Describe the bug**
Giving xp as a GM (targeting a character token) generates an error on the console:

foundry.js:319 Error: Syntax error, unrecognized expression: Received <b>100</b> experience for <b>Just Because Testing 1</b>
    at Function.se.error (jquery.min.js:2)
    at se.tokenize (jquery.min.js:2)
    at se.select (jquery.min.js:2)
    at Function.se [as find] (jquery.min.js:2)
    at S.fn.init.find (jquery.min.js:2)
    at new S.fn.init (jquery.min.js:2)
    at S (jquery.min.js:2)
    at chat.js:288
    at Function._call (foundry.js:294)
    at Function.callAll (foundry.js:253)

**To Reproduce**
Write  /exp 123 Reason in chat as GM, select a player token and click to receive to give xp

![image](https://user-images.githubusercontent.com/17179777/158017002-28394a2e-8881-4123-b69a-012e92de4305.png)


PLEASE PROVIDE YOUR FOUNDRY, SYSTEM, and MODULE VERSIONS (where applicable) 

Foundry: 9 255
wfrp4e: 5.1.1
wfrp4e-core: 2.4.1
wfrp4e-starter-set: 2.3.1
wfrp4e-rnhd: 2.4.1
wfrp4e-eis:
wfrp4e-ua1:
wfrp4e-dotr:
wfrp4e-archives1:
wfrp4e-middenheim:

--- a/README.md
+++ b/README.md
@@ -1,7 +1,7 @@
 # Warhammer Fantasy Roleplay 4th Ed. (FoundryVTT)
 
 ![](https://user-images.githubusercontent.com/28637157/97379891-e4a1cc00-1893-11eb-9e0c-d93b92844d5b.jpg)
-**[Current Version]**: `5.1.1`
+**[Current Version]**: `5.2.0`
 
 **[Compatibility]**: `FoundryVTT V9`
 

--- a/css/wfrp4e.css
+++ b/css/wfrp4e.css
@@ -15823,6 +15823,22 @@ input[name="dd-import.importPath"] {
   /* background: #ff000047; */
 }
 
+#world-config .form-group-stacked {
+  background: url(../ui/frames/parchment-texture.webp) repeat;
+  border: 2px groove #564d3a;
+  border-image: url(../ui/chat-controls-border.webp) 4 repeat;
+  border-image-width: 4px;
+  border-image-outset: 1px;
+}
+
+#world-config .form-group-stacked .world-desc {
+  font-family: var(--actor-input-font-family);
+  font-size: var(--actor-input-font-size);
+  font-weight: var(--actor-input-font-weight);
+  color: var(--actor-input-color);
+  text-align: center;
+}
+
 
 /* FX MASTER STYLING */
 .fxmaster .config.collapsible .flexrow strong {

--- a/lang/en.json
+++ b/lang/en.json
@@ -886,6 +886,7 @@
     "ITEM.ExtraOvercasting" : "Extra Overcast Option",
     "Label" : "Label",
     "ITEM.SkillOverride" : "Skill Override",
+    "ITEM.CastSkillOverride" : "Cast Skill Override",
     "ITEM.AttackType" : "Attack Type",
     "CHAR.WS": "Weapon Skill",
     "CHAR.BS": "Ballistic Skill",

--- a/lang/en.json
+++ b/lang/en.json
@@ -861,7 +861,7 @@
     "WFRP4E.ConditionName.Unconscious" : "Unconscious",
     "WFRP4E.ConditionName.Grappling" : "Grappling",
     "WFRP4E.ConditionName.Fear" : "Fear",
-    "WFRP4E.ConditionName.Defeated" : "Defeated",
+    "WFRP4E.ConditionName.Dead" : "Dead",
 
     "CONDITION.Apply" : "Apply {condition}",
     "CONDITION.ApplyError" : "You cannot apply condition effects to this actor",

--- a/modules/actor/actor-wfrp4e.js
+++ b/modules/actor/actor-wfrp4e.js
@@ -3042,8 +3042,8 @@ export default class ActorWfrp4e extends Actor {
     }
 
     // Revert previous test if rerolled
-    if (test.context.reroll) {
-      let previousFailed = test.context.previousResult.result == "failure"
+    if (test.context.reroll || test.context.fortuneUsedAddSL) {
+      let previousFailed = test.context.previousResult.outcome == "failure"
       switch (strength) {
         case "minor":
           if (previousFailed)
@@ -3070,7 +3070,7 @@ export default class ActorWfrp4e extends Actor {
     let newCorruption = Number(this.status.corruption.value) + corruption
     if (newCorruption < 0) newCorruption = 0
 
-    if (!test.context.reroll)
+    if (!test.context.reroll && !test.context.fortuneUsedAddSL)
       ChatMessage.create(WFRP_Utility.chatDataSetup(game.i18n.format("CHAT.CorruptionFail", { name: this.name, number: corruption }), "gmroll", false))
     else
       ChatMessage.create(WFRP_Utility.chatDataSetup(game.i18n.format("CHAT.CorruptionReroll", { name: this.name, number: corruption }), "gmroll", false))

--- a/modules/actor/sheet/actor-sheet.js
+++ b/modules/actor/sheet/actor-sheet.js
@@ -1473,9 +1473,7 @@ export default class ActorSheetWfrp4e extends ActorSheet {
   }
   _onSystemEffectChanged(ev) {
     let ef = ev.target.value;
-    let data = ev.target.options[ev.target.selectedIndex].dataset
-    let effect = game.wfrp4e.config[data.source][ef]
-    this.actor.createEmbeddedDocuments("ActiveEffect", [effect])
+    this.actor.addSystemEffect(ef)
   }
 
   _onMoneyIconClicked(ev) {

--- a/modules/apps/bug-report.js
+++ b/modules/apps/bug-report.js
@@ -16,7 +16,8 @@ export default class BugReportFormWfrp4e extends Application {
             "Middenheim: City of the White Wolf",
             "Archives of the Empire: Vol 1.",
             "Power Behind the Throne",
-            "Altdorf: Crown of the Empire"
+            "Altdorf: Crown of the Empire",
+            "Ubersreik Adventures II"
         ]
 
         this.domainKeys = [
@@ -30,7 +31,8 @@ export default class BugReportFormWfrp4e extends Application {
             "wfrp4e-middenheim",
             "wfrp4e-archives1",
             "wfrp4e-pbtt",
-            "wfrp4e-altdorf"
+            "wfrp4e-altdorf",
+            "wfrp4e-ua2"
         ]
 
         this.domainKeysToLabel = {
@@ -44,7 +46,8 @@ export default class BugReportFormWfrp4e extends Application {
             "wfrp4e-middenheim" : "middenheim",
             "wfrp4e-archives1" : "archives",
             "wfrp4e-pbtt" : "pbtt",
-            "wfrp4e-altdorf" : "altdorf"
+            "wfrp4e-altdorf" : "altdorf",
+            "wfrp4e-ua2" : "ua2"
         }
     }
 

--- a/modules/apps/item-properties.js
+++ b/modules/apps/item-properties.js
@@ -20,6 +20,11 @@ export default class ItemProperties extends FormApplication {
             this.qualities = foundry.utils.deepClone(game.wfrp4e.config.armorQualities)
             this.flaws = foundry.utils.deepClone(game.wfrp4e.config.armorFlaws)
         }
+        else if (this.object.type == "trapping")
+        {
+            this.qualities = {}
+            this.flaws = {}
+        }
         mergeObject(this.qualities, game.wfrp4e.config.itemQualities)
         mergeObject(this.flaws, game.wfrp4e.config.itemFlaws)
     }

--- a/modules/hooks/chat.js
+++ b/modules/hooks/chat.js
@@ -276,39 +276,6 @@ export default function() {
     }
   });
 
-
-
-  /**
-   * This hook is only used to color code the winner/loser of an opposed test
-   * When the result card shows up, take the start message and apply classes to it
-   */
-  Hooks.on("createChatMessage", (msg, options) => {
-
-    // If message has the opposed class signifying an opposed result
-    if ($(msg.data.content).find(".opposed-card").length && msg.data.flags.startMessageId && (game.user.isUniqueGM)) {
-      // Look in the flags for the winner and startMessage
-      let winner = msg.data.flags.opposeData.opposeResult.winner;
-      let startMessage = game.messages.get(msg.data.flags.startMessageId)
-      // The loser is "attacker" or "defender"
-      let loser = winner == "attacker" ? "defender" : "attacker"
-      // forgive me but i'm too tired to deal with jquery
-
-      // Replace "attacker" with "attacker winner" or "defender" with "defender winner" to apply the color coded borders
-      let newContent = startMessage.data.content.replace(winner, `${winner} winner`)
-      newContent = newContent.replace(loser, `${loser} loser`)
-
-      // Update card with new content
-      let cardData = {
-        user: game.user.id,
-        content: newContent
-      }
-      startMessage.update(cardData).then(resultCard => {
-        ui.chat.updateMessage(resultCard)
-      })
-    }
-  });
-
-
   /**
  * Searches each message and adds drag and drop functionality and hides certain things from players
  */

--- a/modules/hooks/chat.js
+++ b/modules/hooks/chat.js
@@ -31,7 +31,6 @@ export default function() {
     let rollMode = game.settings.get("core", "rollMode");
     if (["gmroll", "blindroll"].includes(rollMode)) msg["whisper"] = ChatMessage.getWhisperRecipients("GM").map(u => u.id);
     if (rollMode === "blindroll") msg["blind"] = true;
-    msg["type"] = 0;
 
     let regExp;
     regExp = /(\S+)/g;

--- a/modules/item/item-wfrp4e.js
+++ b/modules/item/item-wfrp4e.js
@@ -136,6 +136,9 @@ export default class ItemWfrp4e extends Item {
 
 
       if (this.encumbrance && this.quantity) {
+        if (this.properties?.qualities?.lightweight)
+          this.encumbrance.value -= 1
+
         this.encumbrance.value = (this.encumbrance.value * this.quantity.value)
         if (this.encumbrance.value % 1 != 0)
           this.encumbrance.value = this.encumbrance.value.toFixed(2)
@@ -410,6 +413,11 @@ export default class ItemWfrp4e extends Item {
   _trappingExpandData() {
     let data = this.toObject().data
     data.properties = [];
+
+    let itemProperties = this.Qualities.concat(this.Flaws)
+    for (let prop of itemProperties)
+      data.properties.push("<a class ='item-property'>" + prop + "</a>")
+
     return data;
   }
 
@@ -559,10 +567,9 @@ export default class ItemWfrp4e extends Item {
     if (this.damageToItem.shield)
       properties.push(`${game.i18n.localize("ITEM.ShieldDamaged")} ${this.damageToItem.shield} points`)
 
-    let weaponProperties = this.qualities.value.map(i => game.wfrp4e.utility.qualityList[i.name] + " " + i.value).concat(this.flaws.value.map(i => game.wfrp4e.utility.flawList[i.name] + " " + i.value));
-    for (let prop in weaponProperties)
-      if (Array.isArray(weaponProperties[prop]))
-        properties = properties.concat(weaponProperties[prop].map(i => i = "<a class ='item-property'>" + i + "</a>"))
+      let itemProperties = this.OriginalQualities.concat(this.OriginalFlaws)
+      for (let prop of itemProperties)
+        properties.push("<a class ='item-property'>" + prop + "</a>")
 
     if (this.special.value)
       properties.push(`${game.i18n.localize("Special")}: ` + this.special.value);
@@ -576,10 +583,9 @@ export default class ItemWfrp4e extends Item {
     let data = this.toObject().data
     let properties = [];
     properties.push(game.wfrp4e.config.armorTypes[this.armorType.value]);
-    let armourProperties = this.qualities.value.map(i => game.wfrp4e.utility.qualityList[i.name] + " " + i.value).concat(this.flaws.value.map(i => game.wfrp4e.utility.flawList[i.name] + " " + i.value));
-    for (let prop in armourProperties)
-      if (Array.isArray(armourProperties[prop]))
-        properties = properties.concat(armourProperties[prop].map(i => i = "<a class ='item-property'>" + i + "</a>"))
+    let itemProperties = this.Qualities.concat(this.Flaws)
+    for (let prop of itemProperties)
+      properties.push("<a class ='item-property'>" + prop + "</a>")
     properties.push(this.penalty.value);
 
     data.properties = properties.filter(p => !!p);
@@ -602,10 +608,9 @@ export default class ItemWfrp4e extends Item {
       properties.push(`${game.i18n.localize("Damage")}: ${damage}`);
     }
 
-    let ammoProperties = this.qualities.value.map(i => game.wfrp4e.utility.qualityList[i.name] + " " + i.value).concat(this.flaws.value.map(i => game.wfrp4e.utility.flawList[i.name] + " " + i.value));
-    for (let prop in ammoProperties)
-      if (Array.isArray(ammoProperties[prop]))
-        properties = properties.concat(ammoProperties[prop].map(i => i = "<a class ='item-property'>" + i + "</a>"))
+    let itemProperties = this.Qualities.concat(this.Flaws)
+    for (let prop of itemProperties)
+      properties.push("<a class ='item-property'>" + prop + "</a>")
 
     if (this.special.value)
       properties.push(`${game.i18n.localize("Special")}: ` + this.special.value);
@@ -771,6 +776,14 @@ export default class ItemWfrp4e extends Item {
       `<b>${game.i18n.localize("Encumbrance")}</b>: ${this.encumbrance.value}`,
       `<b>${game.i18n.localize("Availability")}</b>: ${game.wfrp4e.config.availability[this.availability.value] || "-"}`
     ]
+
+    // Make qualities and flaws clickable
+    if (this.qualities.value.length)
+      properties.push(`<b>${game.i18n.localize("Qualities")}</b>: ${this.OriginalQualities.map(i => i = "<a class ='item-property'>" + i + "</a>").join(", ")}`);
+
+    if (this.flaws.value.length)
+      properties.push(`<b>${game.i18n.localize("Flaws")}</b>: ${this.OriginalFlaws.map(i => i = "<a class ='item-property'>" + i + "</a>").join(", ")}`);
+
     return properties;
   }
 
@@ -1645,6 +1658,11 @@ export default class ItemWfrp4e extends Item {
 
   get properties() {
 
+    if (!this.qualities || !this.flaws)
+    {
+      return {}
+    }
+
     let properties = {
       qualities: ItemWfrp4e._propertyArrayToObject(this.qualities.value, game.wfrp4e.utility.qualityList()),
       flaws: ItemWfrp4e._propertyArrayToObject(this.flaws.value, game.wfrp4e.utility.flawList()),
@@ -1658,9 +1676,7 @@ export default class ItemWfrp4e extends Item {
         properties.qualities = this.ammo.properties.qualities
     }
 
-
-
-    properties.special = this.special.value
+    properties.special = this.special?.value
     if (this.ammo)
       properties.specialAmmo = this.ammo.properties.special
 

--- a/modules/item/item-wfrp4e.js
+++ b/modules/item/item-wfrp4e.js
@@ -1599,7 +1599,17 @@ export default class ItemWfrp4e extends Item {
         skill = skills.find(x => x.name.toLowerCase().includes(`(${this.WeaponGroup.toLowerCase()})`))
     }
     if (this.type == "spell")
-      skill = skills.find(i => i.name.toLowerCase() == `${game.i18n.localize("NAME.Language")} (${game.i18n.localize("SPEC.Magick")})`.toLowerCase())
+    {
+      // Use skill override, if not found, use Language (Magick)
+      if (this.skill.value)
+      {
+        skill = skills.find(i => i.name.toLowerCase() == this.skill.value.toLowerCase())
+      }
+      if (!skill)
+      {
+        skill = skills.find(i => i.name.toLowerCase() == `${game.i18n.localize("NAME.Language")} (${game.i18n.localize("SPEC.Magick")})`.toLowerCase())
+      }
+    }
 
     if (this.type == "prayer")
       skill = skills.find(i => i.name.toLowerCase() == game.i18n.localize("NAME.Pray").toLowerCase())

--- a/modules/system/opposed-wfrp4e.js
+++ b/modules/system/opposed-wfrp4e.js
@@ -110,6 +110,7 @@ export default class OpposedWFRP {
     await this.opposedTest.evaluate();
     this.formatOpposedResult();
     this.renderOpposedResult()
+    this.colorWinnerAndLoser()
   }
 
   renderOpposedStart() {
@@ -229,6 +230,26 @@ export default class OpposedWFRP {
     }
 
     return opposeResult;
+  }
+
+  colorWinnerAndLoser() 
+  {
+    try {
+      let winner = this.opposedTest.opposeResult.winner
+
+      // The loser is "attacker" or "defender"
+      let loser = winner == "attacker" ? "defender" : "attacker"
+
+      // Replace "attacker" with "attacker winner" or "defender" with "defender winner" to apply the color coded borders
+      let content = this.message.data.content
+      content = content.replace(winner, `${winner} winner`)
+      content = content.replace(loser, `${loser} loser`)
+      return this.message.update({content})
+    }
+    catch(e)
+    {
+      console.error(`Error color coding winner and loser: ${e}`)
+    }
   }
 
 

--- a/modules/system/rolls/test-wfrp4e.js
+++ b/modules/system/rolls/test-wfrp4e.js
@@ -120,11 +120,13 @@ export default class TestWFRP {
   }
 
   async addSL(SL) {
+    this.context.previousResult = duplicate(this.result)
     this.preData.SL = Math.trunc(this.result.SL) + SL;
     this.preData.slBonus = 0;
     this.preData.successBonus = 0;
     this.preData.roll = Math.trunc(this.result.roll);
-    this.preData.hitloc = this.result.hitloc.roll;
+    if (this.preData.hitLocation)
+      this.preData.hitloc = this.result.hitloc.roll;
 
     this.roll()
   }

--- a/modules/system/rolls/test-wfrp4e.js
+++ b/modules/system/rolls/test-wfrp4e.js
@@ -223,6 +223,11 @@ export default class TestWFRP {
         else
           SL = 0;
         SL += slBonus
+
+        if (Number.isNumeric(this.preData.SL))
+        {
+          SL = this.preData.SL
+        }
       }
       SL += successBonus;
       if (this.result.roll <= 5 && SL < 1 && !this.context.unopposed)

--- a/system.json
+++ b/system.json
@@ -2,7 +2,7 @@
   "name": "wfrp4e",
   "title": "Warhammer Fantasy Roleplay 4th Edition",
   "description": "A comprehensive system for running grim and perilous games of Warhammer Fantasy Roleplay in the Foundry VTT environment.",
-  "version": "5.1.1",
+  "version": "5.2.0",
   "author": "Moo Man, CatoThe1stElder",
   "authors" : [
     {
@@ -53,6 +53,6 @@
   "templateVersion":3,
   "socket": true,
   "manifest" :  "https://raw.githubusercontent.com/moo-man/WFRP4e-FoundryVTT/master/system.json",
-  "download" :  "https://github.com/moo-man/WFRP4e-FoundryVTT/releases/download/5.1.1/wfrp4e.zip",
+  "download" :  "https://github.com/moo-man/WFRP4e-FoundryVTT/releases/download/5.2.0/wfrp4e.zip",
   "url" : "https://github.com/moo-man/WFRP4e-FoundryVTT"
 }

--- a/template.json
+++ b/template.json
@@ -774,6 +774,14 @@
         "type": "Number",
         "label": "Location",
         "value": 0
+      },
+      "qualities": {
+        "label": "Qualities",
+        "value": []
+      },
+      "flaws": {
+        "label": "Flaws",
+        "value": []
       }
     },
     "talent": {

--- a/template.json
+++ b/template.json
@@ -885,6 +885,9 @@
         "label": "Memorized",
         "value": false
       },
+      "skill" : {
+        "value" : ""
+      },
       "ingredients": [],
       "currentIng": {
         "type": "Number",

--- a/templates/items/item-spell-sheet.html
+++ b/templates/items/item-spell-sheet.html
@@ -47,6 +47,13 @@
       </div>
 
       <div class="form-group">
+        <label class="label-text">{{localize "ITEM.CastSkillOverride"}}</label>
+        <div class="input-box">
+          <input class="input-text" type="text" name="data.skill.value" value="{{data.skill.value}}" data-dtype="String" />
+        </div>
+      </div>
+
+      <div class="form-group">
         <label class="label-text">{{localize "Casting Number"}}</label>
         <div class="input-box">
           <input class="input-text" type="text" name="data.cn.value" value="{{data.cn.value}}"

--- a/templates/items/item-trapping-sheet.html
+++ b/templates/items/item-trapping-sheet.html
@@ -35,6 +35,9 @@
           {{/select}}
         </select>
       </div>
+      
+      {{> systems/wfrp4e/templates/partials/qualities-flaws.html}}
+
     </div>
     {{> systems/wfrp4e/templates/items/item-effects.html}}
 

